### ISSUE TITLE
Process memory usage reports process under wrong name after fork

### ISSUE THREAD

**Issue by schnittchen** [appsignal/appsignal-ruby#927]
We have a Rails app on MRI running under Passenger in the "standalone" configuration. This keeps a process around with a preloaded app, to be forked to actual worker processes for serving requests (this is done to save memory due to copy-on-write of memory pages).

The general process is documented here: https://github.com/phusion/passenger/tree/stable-6.0/src/agent/Core/SpawningKit#preloaders

The processes can be distinguished by their name in the process table (visible in tools like `ps` or `htop`).

The `process_rss` metric, which has tags "hostname" and "process_name", only shows three lines currently in the graph, two unrelated (rake was started during the current time period, and sidekiq) and a process named "Passenger AppPreloader: $path - $name": This process looks very much like the mentioned preloader process. It looked to us like there was a memory leak in this process, as the line was basically ever increasing until a crash or graceful termination.

My suspicion is that the instrumentation uses stale values after fork() in the child processes, based on these observations:

* the graph shows no line for any worker process, although there are such processes
* sometimes, the preloader process disappears (I don't know why, but assume that's normal behaviour), yet the graph line for this process proceeds to ascend
* the preloader process should only show continued memory increase under very odd circumstances, since it is not involved in handling requests

Maybe this behaviour is reproducible in a toy program using fork and setting the process name with `Process.setproctitle(new_name)` (I've noticed that htop seems to cache this, so the new name is only visible after restarting htop)

**Comment by tombruijn** [comment]
Hi @schnittchen! Our Ruby gem uses the [`$PROGRAM_NAME` global variable](https://github.com/appsignal/appsignal-ruby/blob/2fc6ba85be1e0cabc2bb8fb26469ad47d1c60243/lib/appsignal/config.rb#L335) to determine the process name we use to report the process memory for. Whatever AppSignal is initialized with on app boot, that's what's used to report the metric. If that changes after AppSignal is started, we don't pick up the change. If a process is forked, it doesn't (re)start AppSignal, so it keeps whatever process name the parent process was named. That's what I think is happening here. Does that fit your theory?

**Comment by schnittchen** [comment]
Hey Tom, that sounds very much like I was thinking, yes.

**Comment by tombruijn** [comment]
All 3 lines are needed.
Including the one where you set `$PROGRAM_NAME = "new process name"`

**Comment by schnittchen** [comment]
I'm doing this in an initializer in a Rails app, and I see all of the logging I put in here:

```
def log_once(tag)
  if Rails.logger
    Rails.logger.warn({tag: tag, pid: $$, program_name: $PROGRAM_NAME}.inspect)
  end 
end

def log_forever(tag)
  Thread.new do
    loop do
      log_once(tag)
      sleep(20)
    end 
  end 
end

log_forever(:top_level)

if defined?(PhusionPassenger)
  PhusionPassenger.on_event(:starting_worker_process) do |forked|
    log_once(:starting_worker_process)

    if forked
      Appsignal.config = nil 
      Appsignal.start
      Appsignal.start_logger
    end 

    log_forever(:after_fork)
  end 
end
```

In particular, I see

```
WARN -- : {:tag=>:starting_worker_process, :pid=>165113, :program_name=>"Passenger RubyApp: /home/app/production/current"}
```

but the `process_rss` dashboard graph still shows only "Passenger AppPreloader" and not "Passenger RubyApp" for the `process_name` tag

**Comment by tombruijn** [comment]
When I reproduce this I see this in the AppSignal logs:

> [2023-03-29T13:23:40 (process) #889][INFO] Starting AppSignal 3.3.2 (Passenger RubyApp: /app (development), Ruby 3.1.2, aarch64-linux)
> [2023-03-29T13:23:40 (extension) #889][Warn] Extension started while it was already running

So the extension doesn't want to start a new one for the forked process, like it did in my [test app posted earlier](https://github.com/appsignal/appsignal-ruby/issues/927#issuecomment-1487313486).

To improve this properly, it looks like we need to make some changes to our gem and/or extension. Not quite sure how yet though. Either read it from wherever htop and friends reads it from, or add a method like `Appsignal.setproctitle`.
I'll discuss it with the team if we want to pick this up and when.

**Comment by tombruijn** [comment]
Suggestion: add the PID to the tag name. The name would still be wrong, but at least the process metrics will be reported uniquely.

**Comment by tombruijn** [comment]
Hi @schnittchen, I've just published Ruby gem 3.4.3 that now also tracks the process_rss metric with another tag combination including the process_id.

Here's a dashboard you can import:

```json
{
  "title": "Process memory usage by id",
  "description": "",
  "visuals": [
    {
      "title": "Average memory usage by process id",
      "description": "Average RSS memory usage of your processes split out per host and process type.",
      "line_label": "%process_id% - %hostname%",
      "display": "LINE",
      "format": "size",
      "format_input": "kilobyte",
      "draw_null_as_zero": true,
      "metrics": [
        {
          "name": "process_rss",
          "fields": [
            {
              "field": "MEAN"
            }
          ],
          "tags": [
            {
              "key": "hostname",
              "value": "*"
            },
            {
              "key": "process_id",
              "value": "*"
            }
          ]
        }
      ],
      "type": "timeseries"
    },
    {
      "title": "95th percentile memory usage by process id",
      "description": "95th percentile RSS memory usage of your processes split out per host and process type.",
      "line_label": "%process_id% %hostname%",
      "display": "LINE",
      "format": "size",
      "format_input": "kilobyte",
      "draw_null_as_zero": true,
      "metrics": [
        {
          "name": "process_rss",
          "fields": [
            {
              "field": "P95"
            }
          ],
          "tags": [
            {
              "key": "hostname",
              "value": "*"
            },
            {
              "key": "process_id",
              "value": "*"
            }
          ]
        }
      ],
      "type": "timeseries"
    }
  ]
}
```

Closing this for now.
